app/src/main/java/org/mozilla/focus/utils/UrlUtils.java

```diff
@@ -82,38 +84,32 @@ public static String getSearchTermsOrUrl(Context context, String url) {
     }
 
     /**
-     * @return If a url matches the form of a first search URL for the current default search engine.
+     * @return The detected search engine or null if one was not matched with the URL
      */
-    private static boolean isDefaultSearchUrl(Context context, String urlSearch) {
+    private static SearchEngine isDefaultSearchUrl(String urlSearch) {
         if (TextUtils.isEmpty(urlSearch)) {
-            return false;
+            return null;
         }
 
-        final SearchEngine searchEngine = SearchEngineManager.getInstance()
-                .getDefaultSearchEngine(context);
-
-        final String baseSearchUrl = searchEngine.getBaseSearchUrl();
-        if (TextUtils.isEmpty(baseSearchUrl)) {
-            return false;
-        }
-        // %7B is a placeholder for search terms - if it is a default search engine url, the urls should match before the placeholder.
-        if (baseSearchUrl.contains("%7B")) {
-            final int searchPlaceholderIndex = baseSearchUrl.lastIndexOf("%7B");
-            return urlSearch.startsWith(baseSearchUrl.substring(0, searchPlaceholderIndex));
-        } else {
-            return false;
+        final String representativeSnippet = UrlUtils.getRepresentativeSnippet(urlSearch);
+        final List<SearchEngine> searchEngines = SearchEngineManager.getInstance().getSearchEngines();
+        for (SearchEngine searchEngine : searchEngines) {
+            if (representativeSnippet.contains(searchEngine.getName().toLowerCase())) {
+                return searchEngine;
+            }
         }
+        return null;
     }
 
     /**
      * @return The search parameters decoded from the search url or null if they could not be extracted.
      */
-    private static String getSearchTermsFromUrl(Context context, String url) {
+    private static String getSearchTermsFromUrl(String url, SearchEngine searchEngine) {
         String searchParams = null;
         final Uri uri = Uri.parse(url);
         try {
-            final String queryKey = getSearchQueryKeyForSearchEngine(context);
-            if (queryKey != null) {
+            final String queryKey = searchEngine.getSearchTermsParamName();
+            if (queryKey != null && url.contains(queryKey)) {
```

Comment:
Afaik the param name for Google is just "q". It's likely that there's a URL that contains google.com and a "q". You already have an `Uri` object here, you might be able to ask it if there's an actual query parameter with the name "q". But nevertheless it could happen that there are URLs with google.com and a "q" parameter that are not actually search result pages. :(